In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv("FoodBalanceSheets_E_Africa_NOFLAG.csv", encoding = "latin-1")
df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

#### Question 1

In [3]:
df.shape

(60943, 12)

#### Question 2
What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [4]:
round(df["Y2017"].describe(), 2)

count     59437.00
mean        140.92
std        1671.86
min       -1582.00
25%           0.00
50%           0.10
75%           9.00
max      190873.00
Name: Y2017, dtype: float64

#### Question 3
Given the following numpy array 

array  = ([[94, 89, 63],

          [93, 92, 48],
          [92, 94, 56]])

How would you select  the elements in bold and italics from the array?


In [5]:
array  = np.array([[94, 89, 63],

             [93, 92, 48],

             [92, 94, 56]])

array[:2, 1:]

array([[89, 63],
       [92, 48]])

#### Question 5
Answer the following questions based on the African food production dataset provided by the FAO website already provided

What is the total sum of Wine produced in 2015 and 2018 respectively?

In [6]:
df.groupby("Item")[["Y2015", "Y2018"]].sum()

Y2015      Y2018
Item                                     
Alcohol, Non-Food      2180.00    2293.00
Alcoholic Beverages   98783.72   97847.27
Animal Products       11811.73   11578.61
Animal fats          200675.72  269648.27
Apples and products   10559.15    9640.51
...                        ...        ...
Vegetables, Other    158104.08  163987.21
Vegetal Products     107064.17  107775.39
Wheat and products   234710.51  242645.19
Wine                   4251.81    4039.32
Yams                 203151.78  221272.09

[119 rows x 2 columns]

In [7]:
# to confirm

wine = df[df["Item"] == "Wine"]

wine.groupby("Item")[["Y2015", "Y2018"]].sum()

Y2015    Y2018
Item                  
Wine  4251.81  4039.32

> Total sum of wine produced in 2015 and 2018 respectively are 4251.81 and	4039.32

#### Question 6
Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?

In [8]:
df.groupby("Area")["Y2017"].sum().sort_values(ascending = False)

Area
Nigeria                        1483268.23
Egypt                           866379.92
South Africa                    517590.54
Ethiopia                        448683.76
Morocco                         388495.36
Ghana                           337599.06
Algeria                         325644.27
United Republic of Tanzania     322616.85
Kenya                           264660.66
Sudan                           239931.92
Cameroon                        232030.43
Angola                          229159.57
Côte d'Ivoire                   224599.01
Uganda                          213950.38
Malawi                          181098.71
Mozambique                      161407.98
Mauritania                      156665.46
Mali                            149928.33
Madagascar                      131197.73
Niger                           126707.58
Benin                           124771.22
Tunisia                         124167.20
Zambia                          103223.77
Burkina Faso                 

> Nigeria has the highest sum with 1483268.23

#### Question 7
Consider the following list of tuples:

y = [(2, 4), (7, 8), (1, 5, 9)]

How would you assign element 8 from the list to a variable x?

In [9]:
y = [(2, 4), (7, 8), (1, 5, 9)]
x = y[1][1]
x

8

#### Question 8
What is the total number of unique countries in the dataset?

In [10]:
# to get the number of unique countries
df["Area"].nunique()

49

#### Question 10
Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?

In [11]:
df.groupby("Area")["Y2017"].sum().sort_values().head(7)

Area
Sudan (former)               0.00
Ethiopia PDR                 0.00
Comoros                     59.84
Seychelles                 442.34
Sao Tome and Principe    12662.63
Cabo Verde               14650.74
Guinea-Bissau            19102.77
Name: Y2017, dtype: float64

> Guinea-Bissau has the 7th lowest sum in 2017

#### Question 11
Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?

In [12]:
df.groupby("Element")[["Y2014", "Y2015", "Y2016", "Y2017", "Y2018"]].sum()

Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein supply quantity (g/capita/day)    11842.45    11833.56  
Residuals                                 35500.00    34864.00  
Seed                                      24870.14    25263.14  
Stock Variation                           54316.91    20577.91  
Total Population - Both sexes           1112641.00  1140605.00  
Tourist consumption                          91.00       90.00

In [13]:
# to confirm

stock_var =  df[df["Element"] == "Stock Variation"]
stock_var.groupby("Element")[["Y2014", "Y2015", "Y2016", "Y2017", "Y2018"]
                            ].sum().transpose().sort_values(by = "Stock Variation", ascending = False)

Element  Stock Variation
Y2014           58749.83
Y2017           54316.91
Y2015           34910.99
Y2016           33140.12
Y2018           20577.91

#### Question 12
Given the following python code, what would the output of the code give?

my_tuppy = (1,2,5,8)

my_tuppy[2] = 6

In [14]:
my_tuppy = (1,2,5,8)

my_tuppy[2] = 6

TypeError: 'tuple' object does not support item assignment

#### Question 14
If you have the following list

lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]

col = [‘Age’,’Nationality’,’Overall’]

How do you create a pandas DataFrame using this list, to look like the table below?

In [15]:
lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]

col = ["Age","Nationality","Overall"]

df1 = pd.DataFrame(lst, columns = col, index = [i for i in range(1,4)])
df1

Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

#### Question 16
What is the total number and percentage of missing data in 2014 to 3 decimal places?

In [16]:
# total number of missing data in 2014

df["Y2014"].isnull().sum()

1589

In [17]:
# percentage of missing data in 2014 to 3dp

round(df["Y2014"].isnull().sum()* 100 / len(df), 3)

2.607

#### Question 17
Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?

In [18]:
df.groupby("Element")["Y2017"].sum()

Element
Domestic supply quantity                  2088198.10
Export Quantity                            182338.80
Fat supply quantity (g/capita/day)          10253.84
Feed                                       223705.68
Food                                      1258888.28
Food supply (kcal/capita/day)              454681.00
Food supply quantity (kg/capita/yr)         48690.04
Import Quantity                            294559.09
Losses                                     160614.00
Other uses (non-food)                       91645.97
Processing                                 292836.00
Production                                2030056.89
Protein supply quantity (g/capita/day)      11842.45
Residuals                                   35500.00
Seed                                        24870.14
Stock Variation                             54316.91
Total Population - Both sexes             1112641.00
Tourist consumption                            91.00
Name: Y2017, dtype: float64

In [19]:
# more specifically

df.groupby("Element")["Y2017"].sum()["Processing"]

292836.0

#### Question 18
What is the total Protein supply quantity in Madagascar in 2015?

In [20]:
mad =  df[df["Area"] == "Madagascar"]

In [21]:
mad.groupby("Element")["Y2015"].sum()["Protein supply quantity (g/capita/day)"]

173.05

#### Question 19
What would be the output for?

S = [['him', 'sell'], [90, 28, 43]]

S[0][1][1]

In [22]:
S = [['him', 'sell'], [90, 28, 43]]

S[0][1][1]

'e'

#### Question 20
Which year had the least correlation with ‘Element Code’?

In [23]:
df.corr()

Area Code  Item Code  Element Code     Y2014     Y2015  \
Area Code      1.000000  -0.005159     -0.000209  0.006164  0.005472   
Item Code     -0.005159   1.000000     -0.024683  0.021722  0.020857   
Element Code  -0.000209  -0.024683      1.000000  0.024457  0.023889   
Y2014          0.006164   0.021722      0.024457  1.000000  0.994647   
Y2015          0.005472   0.020857      0.023889  0.994647  1.000000   
Y2016          0.005247   0.020109      0.023444  0.996081  0.995739   
Y2017          0.005006   0.021494      0.024254  0.995230  0.988048   
Y2018          0.005665   0.021314      0.024279  0.994872  0.988208   

                 Y2016     Y2017     Y2018  
Area Code     0.005247  0.005006  0.005665  
Item Code     0.020109  0.021494  0.021314  
Element Code  0.023444  0.024254  0.024279  
Y2014         0.996081  0.995230  0.994872  
Y2015         0.995739  0.988048  0.988208  
Y2016         1.000000  0.992785  0.992757  
Y2017         0.992785  1.000000  0.998103  
Y2018         0.992757  0.998103  1.000000

In [24]:
# more specifically

df.corr()["Element Code"]

Area Code      -0.000209
Item Code      -0.024683
Element Code    1.000000
Y2014           0.024457
Y2015           0.023889
Y2016           0.023444
Y2017           0.024254
Y2018           0.024279
Name: Element Code, dtype: float64